### PR TITLE
Add multi-arch native image and Mandrel image push steps

### LINKED ISSUE #83: Provide multi-arch images
Trying to use image on aarch64 and it's not working:

```
[INFO] [io.quarkus.deployment.pkg.steps.NativeImageBuildStep] podman run -v /home/g/downloads/code-with-quarkus/target/code-with-quarkus-1.0.0-SNAPSHOT-native-image-source-jar:/project:z --env LANG=C --userns=keep-id --rm quay.io/quarkus/ubi-quarkus-native-image:20.1.0-java11 -J-Dsun.nio.ch.maxUpdateArraySize=100 -J-Djava.util.logging.manager=org.jboss.logmanager.LogManager -J-Dvertx.logger-delegate-factory-class-name=io.quarkus.vertx.core.runtime.VertxLogDelegateFactory -J-Dvertx.disableDnsResolver=true -J-Dio.netty.leakDetection.level=DISABLED -J-Dio.netty.allocator.maxOrder=1 -J-Duser.language=en -J-Dfile.encoding=UTF-8 --initialize-at-build-time= -H:InitialCollectionPolicy=com.oracle.svm.core.genscavenge.CollectionPolicy$BySpaceAndTime -H:+JNI -jar code-with-quarkus-1.0.0-SNAPSHOT-runner.jar -H:FallbackThreshold=0 -H:+ReportExceptionStackTraces -H:-AddAllCharsets -H:-IncludeAllTimeZones -H:EnableURLProtocols=http --no-server -H:-UseServiceLoaderFeature -H:+StackTrace code-with-quarkus-1.0.0-SNAPSHOT-runner
{"msg":"exec container process `/opt/graalvm/bin/native-image`: Exec format error","level":"error","time":"2020-07-29T10:03:36.000781782Z"}
```

The image descriptor is ~wrongly~ assuming all images are for `amd64`, see [here](https://github.com/quarkusio/quarkus-images/blob/master/modules/graalvm/20.1.0-java11/module.yaml#L9).

Since GraalVM offers aarch64 downloads, should we provode complementary images for that arch?

### PR DESCRIPTION
**Pull request by galderz**
This should close #83, but I think we'll only know if it really works once integrated and we try to push the images. @cescoffier?

/cc @zakkak Clement had suggested that maybe push images script should become a jbang script. WDYT?

### CODE REVIEW

File: .github/push-images.sh [hunk 1]
```diff
@@ -10,3 +10,39 @@ do
...
+if [[ $(docker images | grep quay.io/quarkus/ | grep "ubi-quarkus-mandrel") ]]; then
+    manifest_push "ubi-quarkus-mandrel" "22.0.0.2.Final-java17"
+    manifest_push "ubi-quarkus-mandrel" "22.0.0.2.Final-java11"
+    manifest_push "ubi-quarkus-mandrel" "22.0-java17"
+    manifest_push "ubi-quarkus-mandrel" "22.0-java11"
+
+    manifest_push "ubi-quarkus-mandrel" "22.1.0.0.Final-java17"
+    manifest_push "ubi-quarkus-mandrel" "22.1.0.0.Final-java11"
+    manifest_push "ubi-quarkus-mandrel" "22.1-java17"
+    manifest_push "ubi-quarkus-mandrel" "22.1-java11"
```
Comment (zakkak) [inline .github/push-images.sh:47]: We should definitely find a better way to figure out the tags, ideally fetching them from https://github.com/quarkusio/quarkus-images/blob/54076338ebb857c26b5c8477b11562355aaa59b3/.github/mandrel-images.yaml#L4

If jbang is the preferred way I am OK with that. After all I never had to touch "push-images" before, so I would prefer the people actually maintaining it to decide what they prefer.

Comment (galderz) [reply .github/push-images.sh:47]: When I first discussed this with @cescoffier, I agreed with your statement: let's make this a jbang script and consume the versions in yaml, but I don't think it'll really solve the little hardcoding done in the bash script.

Put it this way: none of the image versions that need to be created (e.g. mandrel `22.1.0.0.Final-java17` or graalvm `22.0.0-java17`) are part of the `versions` that you pass in in the yaml. The versions that you pass in are the ones you build the traditional way (same for build and push) and then you have these extra overarching versions that are only used by push and created and pushed differently (e.g. `docker manifest create...` and `docker manifest push`).

So, I'm currently leaning towards keeping things as they are in the PR. One thing that complicates making wide ranging changes in the push side is that there's no real way to test it until you get the GH action to kick in and see if it works as expected. Note that I actually tested something very similar to this in my own repo's `main` branch and a custom GH action. That's how I ended up pushing multi-arch images to my own fork, see [here](https://quay.io/repository/gzamarre/ubi-quarkus-native-image?tab=tags).

Comment (galderz) [reply .github/push-images.sh:47]: It's true as well that you could enhance the yaml to somehow add the multi arch tags (something akin to what happens with `tags`) and then have some kind of way of saying that this multi-arch tag is built out of these other tags. E.g.

```
multi-arch-tags:
  - id: 22.1-java17-amd64
    manifest-versions:
    - 22.1-java17-amd64
    - 22.1-java17-arm64
  - id: 22.1.0-java17-amd64
    manifest-versions:
    - 22.1.0-java17-amd64
    - 22.1.0-java17-arm64
```

**Comment by edeandrea** [comment]
@galderz I'm happy to test this with the [Quarkus superheroes](https://github.com/quarkusio/quarkus-super-heroes) once there are multi-arch images available somewhere. As I mentioned on #83 I already have the scaffolding & wiring in place. All I need are  `quay.io/quarkus/ubi-quarkus-mandrel` and `quay.io/quarkus/quarkus-micro-image` images that have both amd64 & arm64 platforms available.

**Comment by galderz** [comment]
@edeandrea No work has been done here to add multi arch support for `quarkus-micro-image`. This PR and https://github.com/quarkusio/quarkus-images/pull/195 focused on adding multi-arch support for GraalVM and Mandrel container images.

**Comment by edeandrea** [comment]
Thanks @galderz . Does that mean in the interim I'd have to switch back to the `Dockerfile` that uses `FROM registry.access.redhat.com/ubi8/ubi-minimal:8.5`?

**Comment by galderz** [comment]
> Thanks @galderz . Does that mean in the interim I'd have to switch back to the Dockerfile that uses FROM registry.access.redhat.com/ubi8/ubi-minimal:8.5?

Switch what?

**Comment by edeandrea** [comment]
A `Dockerfile` that uses the `quarkus-micro-image` as it's base image (like https://github.com/quarkusio/quarkus-super-heroes/blob/main/rest-villains/src/main/docker/Dockerfile.native#L20)

**Comment by galderz** [comment]
Hmmmm, yeah, I guess you could change `FROM` to `registry.access.redhat.com/ubi8/ubi-minimal:8.5` and then do things manually yourself if needed. I don't know anything about `quarkus-micro-image` TBH.

**Comment by edeandrea** [comment]
I'll do that once your work on the ubi-mandrel image is done and those images are available.

**Comment by galderz** [comment]
This work won't be integrated as is. The partial work in master is not fully working as expected. An effort is ongoing to provide a better solution.

**Pull request closed without merging**